src/libmongoc/tests/unified/runner.c

```diff
@@ -1329,49 +1392,107 @@ test_check_expected_events_for_client (test_t *test, bson_t *expected_events_for
-      if (too_few_events || too_many_events) {
-         test_set_error (
-            error, "expected: %" PRIu32 " events but got %" PRIu32, expected_num_events, actual_num_events);
-         goto done;
+      if (skip_cse_list_collections (eiter->serialized)) {
+         actual_listCollections_count++;
       }
    }
-
    eiter = entity->events;
+
    bson_iter_t iter;
+   int expected_listCollections_count = 0;
    BSON_FOREACH (expected_events, iter)
    {
-      while (eiter && !event_matches_eventtype (eiter, event_type)) {
-         eiter = eiter->next;
-      }
       bson_t expected_event;
       bson_iter_bson (&iter, &expected_event);
-      if (!eiter) {
-         test_set_error (error, "could not find event: %s", tmp_json (&expected_event));
-         goto done;
+      BSON_FOREACH (&expected_event, iter)
+      {
+         bson_t event_contents;
+         bson_iter_bson (&iter, &event_contents);
+         if (skip_cse_list_collections (&event_contents)) {
+            expected_listCollections_count++;
+            break;
+         }
       }
-      if (!test_check_event (test, &expected_event, eiter, error)) {
-         test_diagnostics_error_info ("checking for expected event: %s", tmp_json (&expected_event));
-         goto done;
+   }
+
+   int expected_num_events = bson_count_keys (expected_events);
+   int actual_num_events = 0;
+   LL_FOREACH (entity->events, eiter)
+   {
+      if (event_matches_eventtype (eiter, event_type)) {
+         actual_num_events++;
+      }
+   }
+
+   {
+      const int difference =
+         (actual_num_events - actual_listCollections_count) - (expected_num_events - expected_listCollections_count);
+      bool too_many_events = difference > 0;
+      bool too_few_events = difference < 0;
+      if (actual_listCollections_count < expected_listCollections_count) {
+         too_few_events = true;
+      }
+
+      if (expected_num_events != actual_num_events) {
+         if (ignore_extra_events && *ignore_extra_events) {
+            // We can never have too many events
+            too_many_events = false;
+         }
+         if (too_few_events || too_many_events) {
+            test_set_error (
+               error, "expected: %" PRIi32 " events but got %" PRIi32, expected_num_events, actual_num_events);
+            goto done;
+         }
+      }
+
+      eiter = entity->events;
+      BSON_FOREACH (expected_events, iter)
+      {
+         bool matched = false;
+         while (eiter && !event_matches_eventtype (eiter, event_type)) {
+            eiter = eiter->next;
+         }
+         bson_t expected_event;
+         bson_iter_bson (&iter, &expected_event);
+
+         do {
+            if (!eiter) {
+               break;
+            }
+            matched = test_check_event (test, &expected_event, eiter, error);
+            if (matched) {
+               continue;
+            }
+
+            if (ignore_extra_events) {
```

Comment:
Changes here suggest that `ignoreExtraEvents` now ignores events in the middle of a sequence.

From [the spec](https://github.com/mongodb/specifications/blob/63e382ee72e63ea65dfb3dd8eff7af7bb7fbb568/source/unified-test-format/unified-test-format.md#expectedeventsforclient):

> `ignoreExtraEvents`: [...] if true, observed events after all specified events have been matched MUST NOT cause a test failure

I expect that means:

```
Expected to pass:
    Actual Events    : A, B, C
    Expected Events  : A, B

Expected to fail:
    Actual Events    : A, B, C
    Expected Events  : A, C
```